sdk/nodejs/provider/server.ts

```diff
@@ -442,9 +442,12 @@ export async function main(provider: Provider, args: string[]) {
     const uncaughtHandler = (err: Error) => {
         if (!uncaughtErrors.has(err)) {
             uncaughtErrors.add(err);
-            // Use `pulumi.log.error` here to tell the engine there was a fatal error, which should
-            // stop processing subsequent resource operations.
-            log.error(err.stack || err.message || ("" + err));
+
+            // log the first error we see, kill any outstanding work, and exit.
+            log.error(err.stack || err.message || ("" + err)).then(() => {
```

Comment:
How does this interact with `process.on("exit")` at L455?

Reply:
More specifically, the assumptions of the implementation without this change is that an uncaught exception/unhandled promise rejection would eventually cause a program failure after all outstanding work completes. I'm observing that this assumption is not totally correct, and that the outstanding work never completes.

Reply:
Which sounds like a key correctness property is violated somewhere we have not found yet, and remains violated after this fix, which is unfortunate, though of course finding the right place can be hard.